### ISSUE TITLE
Unintended behaviour with `deep_update_sync` and `HiddedReinforcer` in inference_core?

### ISSUE THREAD

**Issue by max810** [hkchengrex/XMem#55]
Hi,

I've been reading the code in detail, and found something that looks like unintended behaviour.

In [inference/inference_core.py#L65](https://github.com/hkchengrex/XMem/blob/main/inference/inference_core.py#L65), we create the new hidden state in the decoder using [`HiddenUpdater`](https://github.com/hkchengrex/XMem/blob/main/model/modules.py#L44) module. Now, if we have our `deep_update_sync` disabled, then our `is_normal_update` is always True, thus we save it in the memory, and then later, if it's a **memory frame** AND a **deep update**, update it with [`HiddedReinforcer`](https://github.com/hkchengrex/XMem/blob/main/model/modules.py#L77) module in [inference/inference_core.py#L103](https://github.com/hkchengrex/XMem/blob/main/inference/inference_core.py#L103).

Which is exactly what is stated in the paper in Fig. 6.
![image](https://user-images.githubusercontent.com/29955120/210797393-a29812ff-2447-4f1f-a59d-a4cf5f0c457e.png).

So for **normal updates** we use `HiddenUpdater`, and for **deep updates** we use **both**.

However, if our `deep_update_sync` is enabled (and [by default it is](https://github.com/hkchengrex/XMem/blob/main/eval.py#L58)), then for **normal updates** nothing changes, but for **deep updates** the `is_normal_update` is always going to be `False`, thus the **hidden value** from the decoder [is going to be ignored](https://github.com/hkchengrex/XMem/blob/main/inference/inference_core.py#L70), and ONLY `HiddenReinforcer` will be applied to the hidden state, which looks like unintended behaviour to me.

To summarize:
- If `deep_update_sync` is enabled, then it's either `is_normal_update` (for non-mem frames) or `is_deep_update` (for mem frames)
- If `deep_update_sync` is disabled, then it's `is_normal_update` (for non-mem frames) and **BOTH** (for mem frames)
- So if `deep_update_sync` is enabled, then it's either `HiddenUpdater` or `HiddenReinforcer` applied to hidden state (unintended behaviour?)
- But if `deep_update_sync` is disabled, then it's either `HiddenUpdater` or **BOTH** applied to hidden state (as described in the paper)

Please let me know if this is a bug in implementation or I am missing something. 

Thank you,

**Comment by max810** [comment]
Also, should the `deep_update_sync` be enabled or disabled for training?


Which one did you use to report the results in the paper?

**Comment by hkchengrex** [comment]
Nice catch.

- Figure 6 is what happens during training (i.e., both GRUs update the hidden state). 
- During testing, the default behavior is not to use the normal update GRU when the deep update GRU will be used. This was initially a bug in our implementation. The performance varies a bit with/without this bug, but we do not observe consistent improvement/degradation with different datasets/random seeds. It is however important to use both updates during training. I think this is because the gap between frames in training is much larger (e.g., 5 frames) than in inference. 
- `deep_update_sync` synchronizes working memory insertion (invocation of the value encoder) with deep update. It does not matter in training because the memory works differently. We always set it to `True` during inference in reported results.

**Comment by max810** [comment]
Got it, thanks for the explanation!
